dagjose/multicodec.go

```diff
@@ -1,11 +1,127 @@
 package dagjose
 
+//go:generate go run ./gen .
+//go:generate go fmt ./
+
 import (
-	dagcbor "github.com/ipld/go-ipld-prime/codec/dagcbor"
-	multicodec "github.com/ipld/go-ipld-prime/multicodec"
+	"errors"
+	"io"
+
+	"github.com/ipfs/go-cid"
+	"github.com/ipld/go-ipld-prime/codec"
+	"github.com/ipld/go-ipld-prime/codec/dagcbor"
+	"github.com/ipld/go-ipld-prime/datamodel"
+	"github.com/ipld/go-ipld-prime/linking/cid"
+	"github.com/ipld/go-ipld-prime/multicodec"
+	"github.com/ipld/go-ipld-prime/schema"
+	"github.com/multiformats/go-multibase"
 )
 
 func init() {
-	multicodec.RegisterDecoder(0x85, dagcbor.Decode)
-	multicodec.RegisterEncoder(0x85, dagcbor.Encode)
+	multicodec.RegisterDecoder(0x85, Decode)
+	multicodec.RegisterEncoder(0x85, Encode)
+}
+
+// Decode deserializes data from the given io.Reader and feeds it into the given datamodel.NodeAssembler. Decode fits
+// the codec.Decoder function interface.
+func Decode(na datamodel.NodeAssembler, r io.Reader) error {
+	// If the passed `NodeAssembler` is not of type `_DecodedJOSE__ReprBuilder`, create and use a
+	// `_DecodedJOSE__ReprBuilder`.
+	joseBuilder, alreadyJose := na.(*_DecodedJOSE__ReprBuilder)
+	if !alreadyJose {
+		joseBuilder = Type.DecodedJOSE__Repr.NewBuilder().(*_DecodedJOSE__ReprBuilder)
+	}
+	// DAG-CBOR is a superset of DAG-JOSE and can be used to decode valid DAG-JOSE objects. Use DAG-CBOR decoding but do
+	// not allow IPLD Links. See: https://specs.ipld.io/block-layer/codecs/dag-jose.html
+	err := dagcbor.DecodeOptions{
+		AllowLinks: false,
+	}.Decode(joseBuilder, r)
```

Comment:
I was going to say that this seems unnecessary (because if you're decoding it into something with a schema, that should already enforce this rule and more, right?)... but then remembered the `"link"` field, which has these special rules about how it should be synthesized.  So, okay, maybe it makes sense to do this.

But it draws attention to the fact we seem to be using the types as both a decoding target, *and* as something we return to people to look at... despite those being logically not the same thing.

I can't tell if this is strange looking but fine in both spec and code, or code worth worrying about but still spec fine; or if this hints towards something in the spec that's weird enough that it makes me want to review the spec again.

Reply:
@rvagg, I'm not sure you have any free time, but if you do and could have a look at how this link property is derived and if that makes sense to you, I'd kinda appreciate a doublecheck.

Reply:
Yeah, I'm quite unsatisfied with the contortions I've had to make to match the behavior from the spec. I'm sure much of that is probably my inability to figure out how best to leverage code generation and the general Node assembly framework.

I just pushed what is likely to be the last significant code update to the main body of the codec that adds support for accepting "flattened" serialization. Wanted to put it up quickly (thought not quite refined yet) so you and @rvagg can look at that too.

I have a feeling that some of the utility methods I've added to force the codec into submission might not pass muster 😅 Please let me know if there are better ways to do what I'm trying to achieve, mainly to _re-assemble_ constructed nodes into a different shape (e.g. from a "flattened" node to a "general" node).

Reply:
`link` property is derived like this:
1. take the bytes in the payload (this should be a CID)
2. encode them as a CID Link (in js land that's a CID instance)
3. put it at the `link` property